### PR TITLE
[sntp] fix for ESP-IDF > 5.0

### PR DESCRIPTION
**Pull request by HeMan**
# What does this implement/fix?

Makes sntp works with ESP-IDF >= 5.0

## Types of changes

- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [x] Other

**Related issue or feature (if applicable):** fixes <link to issue>

**Pull request in [esphome-docs](https://github.com/esphome/esphome-docs) with documentation (if applicable):** esphome/esphome-docs#<esphome-docs PR number goes here>

## Test Environment

- [ ] ESP32
- [x] ESP32 IDF
- [ ] ESP8266
- [ ] RP2040
- [ ] BK72xx
- [ ] RTL87xx

## Example entry for `config.yaml`:

```yaml
# Example config.yaml

```

## Checklist:
  - [x] The code change is tested and works locally.
  - [ ] Tests have been added to verify that the new code works (under `tests/` folder).

If user exposed functionality or configuration variables are added/changed:
  - [ ] Documentation added/updated in [esphome-docs](https://github.com/esphome/esphome-docs).

### CODE REVIEW

**Comment by jesserockz** [comment]
What error were you getting before with idf 5+?

**Comment by HeMan** [comment]
> What error were you getting before with idf 5+?


I get multiple errors, redeclared, undeclared and deprecation.
```
/home/jimmy/.platformio/packages/framework-espidf/components/lwip/include/apps/esp_sntp.h:235:6: error: 'void sntp_setoperatingmode(u8_t)' was declared 'extern' and later 'static' [-fpermissive]
  235 | void sntp_setoperatingmode(u8_t operating_mode)
      |      ^~~~~~~~~~~~~~~~~~~~~
In file included from src/esphome/components/sntp/sntp_component.cpp:5:
/home/jimmy/.platformio/packages/framework-espidf/components/lwip/lwip/src/include/lwip/apps/sntp.h:51:6: note: previous declaration of 'void sntp_setoperatingmode(u8_t)'
   51 | void sntp_setoperatingmode(u8_t operating_mode);
      |      ^~~~~~~~~~~~~~~~~~~~~
/home/jimmy/.platformio/packages/framework-espidf/components/lwip/include/apps/esp_sntp.h:242:1: error: expected unqualified-id before '{' token
  242 | {
      | ^
/home/jimmy/.platformio/packages/framework-espidf/components/lwip/include/apps/esp_sntp.h:249:6: error: 'void sntp_setservername(u8_t, const char*)' was declared 'extern' and later 'static' [-fpermissive]
  249 | void sntp_setservername(u8_t idx, const char *server)
      |      ^~~~~~~~~~~~~~~~~~
/home/jimmy/.platformio/packages/framework-espidf/components/lwip/lwip/src/include/lwip/apps/sntp.h:66:6: note: previous declaration of 'void sntp_setservername(u8_t, const char*)'
   66 | void sntp_setservername(u8_t idx, const char *server);
      |      ^~~~~~~~~~~~~~~~~~
/home/jimmy/.platformio/packages/framework-espidf/components/lwip/include/apps/esp_sntp.h:255:6: error: 'void sntp_init()' was declared 'extern' and later 'static' [-fpermissive]
  255 | void sntp_init(void)
      |      ^~~~~~~~~
/home/jimmy/.platformio/packages/framework-espidf/components/lwip/lwip/src/include/lwip/apps/sntp.h:54:6: note: previous declaration of 'void sntp_init()'
   54 | void sntp_init(void);
      |      ^~~~~~~~~
/home/jimmy/.platformio/packages/framework-espidf/components/lwip/include/apps/esp_sntp.h:261:13: error: 'const char* sntp_getservername(u8_t)' was declared 'extern' and later 'static' [-fpermissive]
  261 | const char *sntp_getservername(u8_t idx)
      |             ^~~~~~~~~~~~~~~~~~
/home/jimmy/.platformio/packages/framework-espidf/components/lwip/lwip/src/include/lwip/apps/sntp.h:67:13: note: previous declaration of 'const char* sntp_getservername(u8_t)'
   67 | const char *sntp_getservername(u8_t idx);
      |             ^~~~~~~~~~~~~~~~~~
/home/jimmy/.platformio/packages/framework-espidf/components/lwip/include/apps/esp_sntp.h:267:18: error: 'const ip_addr_t* sntp_getserver(u8_t)' was declared 'extern' and later 'static' [-fpermissive]
  267 | const ip_addr_t* sntp_getserver(u8_t idx)
      |                  ^~~~~~~~~~~~~~
/home/jimmy/.platformio/packages/framework-espidf/components/lwip/lwip/src/include/lwip/apps/sntp.h:59:18: note: previous declaration of 'const ip_addr_t* sntp_getserver(u8_t)'
   59 | const ip_addr_t* sntp_getserver(u8_t idx);
      |                  ^~~~~~~~~~~~~~
src/esphome/components/sntp/sntp_component.cpp: In member function 'virtual void esphome::sntp::SNTPComponent::setup()':
src/esphome/components/sntp/sntp_component.cpp:34:24: warning: 'void sntp_setoperatingmode(u8_t)' is deprecated: use esp_sntp_setoperatingmode() instead [-Wdeprecated-declarations]
   34 |   sntp_setoperatingmode(SNTP_OPMODE_POLL);
      |   ~~~~~~~~~~~~~~~~~~~~~^~~~~~~~~~~~~~~~~~
/home/jimmy/.platformio/packages/framework-espidf/components/lwip/include/apps/esp_sntp.h:235:6: note: declared here
  235 | void sntp_setoperatingmode(u8_t operating_mode)
      |      ^~~~~~~~~~~~~~~~~~~~~
src/esphome/components/sntp/sntp_component.cpp:40:21: warning: 'void sntp_setservername(u8_t, const char*)' is deprecated: use esp_sntp_setservername() instead [-Wdeprecated-declarations]
   40 |   sntp_setservername(0, strdup(this->server_1_.c_str()));
      |   ~~~~~~~~~~~~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/home/jimmy/.platformio/packages/framework-espidf/components/lwip/include/apps/esp_sntp.h:249:6: note: declared here
  249 | void sntp_setservername(u8_t idx, const char *server)
      |      ^~~~~~~~~~~~~~~~~~
src/esphome/components/sntp/sntp_component.cpp:42:23: warning: 'void sntp_setservername(u8_t, const char*)' is deprecated: use esp_sntp_setservername() instead [-Wdeprecated-declarations]
   42 |     sntp_setservername(1, strdup(this->server_2_.c_str()));
      |     ~~~~~~~~~~~~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/home/jimmy/.platformio/packages/framework-espidf/components/lwip/include/apps/esp_sntp.h:249:6: note: declared here
  249 | void sntp_setservername(u8_t idx, const char *server)
      |      ^~~~~~~~~~~~~~~~~~
src/esphome/components/sntp/sntp_component.cpp:45:23: warning: 'void sntp_setservername(u8_t, const char*)' is deprecated: use esp_sntp_setservername() instead [-Wdeprecated-declarations]
   45 |     sntp_setservername(2, strdup(this->server_3_.c_str()));
      |     ~~~~~~~~~~~~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/home/jimmy/.platformio/packages/framework-espidf/components/lwip/include/apps/esp_sntp.h:249:6: note: declared here
  249 | void sntp_setservername(u8_t idx, const char *server)
      |      ^~~~~~~~~~~~~~~~~~
src/esphome/components/sntp/sntp_component.cpp:51:12: warning: 'void sntp_init()' is deprecated: use esp_sntp_init() instead [-Wdeprecated-declarations]
   51 |   sntp_init();
      |   ~~~~~~~~~^~
/home/jimmy/.platformio/packages/framework-espidf/components/lwip/include/apps/esp_sntp.h:255:6: note: declared here
  255 | void sntp_init(void)
      |      ^~~~~~~~~
Compiling .pioenvs/round/src/esphome/components/time/automation.o
*** [.pioenvs/round/src/esphome/components/sntp/sntp_component.o] Error 1
```

File: esphome/components/sntp/sntp_component.cpp [hunk 1]
```diff
@@ -27,13 +22,12 @@ static const char *const TAG = "sntp";
...
 #ifndef USE_HOST
   ESP_LOGCONFIG(TAG, "Setting up SNTP...");
-#if defined(USE_ESP32) || defined(USE_LIBRETINY)
-  if (sntp_enabled()) {
-    sntp_stop();
+#if defined(USE_ESP_IDF)
+  if (esp_sntp_enabled()) {
+    esp_sntp_stop();
   }
-  sntp_setoperatingmode(SNTP_OPMODE_POLL);
```
Comment (jesserockz) [inline esphome/components/sntp/sntp_component.cpp:34]: Is setting the operating mode not required when using libretiny?

Comment (HeMan) [reply esphome/components/sntp/sntp_component.cpp:34]: Great find, fixed!

File: esphome/components/sntp/sntp_component.cpp [hunk 2]
```diff
@@ -27,14 +22,14 @@ static const char *const TAG = "sntp";
...
-#endif
-#ifdef USE_ESP8266
+  esp_sntp_setoperatingmode(ESP_SNTP_OPMODE_POLL);
+#else
   sntp_stop();
+  sntp_setoperatingmode(SNTP_OPMODE_POLL);
```
Comment (kbx81) [inline esphome/components/sntp/sntp_component.cpp:32]: Unless I'm missing something, this wasn't being called on '8266 before this change...but...I just tested it and it is working just fine...so 🤷🏻‍♂️😅 I think this is slightly cleaner than it was before, anyway.

**Review by kbx81 — APPROVED** [review]
Thanks for fixing this! I ran into it as I was doing testing for #6846.

**Comment by nagyrobi** [comment]
After this patch, I get:
```
Compiling .pioenvs/pak-lvgl-test/src/esphome/components/sntp/sntp_component.o
src/esphome/components/sntp/sntp_component.cpp:9:10: fatal error: lwip/apps/sntp.h: No such file or directory
    9 | #include "lwip/apps/sntp.h"
      |          ^~~~~~~~~~~~~~~~~~
compilation terminated.
*** [.pioenvs/pak-lvgl-test/src/esphome/components/sntp/sntp_component.o] Error 1
```
with `host` [platform](https://deploy-preview-3893--esphome.netlify.app/components/host).

Going one step back with ` git reset --hard 2353b2b5e1e66874a84f5f0e1e2ae0fff5780407` I can build successfully.
@clydebarrow @HeMan @kbx81

**Comment by clydebarrow** [comment]
Yes, the changes did not take into account the existence of the `host:` platform.

**Comment by kbx81** [comment]
We will add a test for the `host` platform so this doesn't happen again.

**Comment by clydebarrow** [comment]
the real problem is that allowing the sntp platform on `host` was a hack since it doesn't actually use sntp (and yes, I was responsible for that hack.) I'm raising another PR to add a proper time implementation for host and remove support for it from sntp. Stand by...

**Comment by clydebarrow** [comment]
See #6854

**Pull request merged**